File: python/1963-minimum-number-of-swaps-to-make-the-string-balanced.py
class Solution:
    def minSwaps(self, s: str) -> int:
        extraClose, maxClose = 0, 0

        for c in s:
            if c == "[":
                extraClose -= 1
            else:
                extraClose += 1

            maxClose = max(maxClose, extraClose)

        return (maxClose + 1) // 2  # Or math.ceil(maxClose / 2)

File: rust/1963-minimum-number-of-swaps-to-make-the-string-balanced.rs
impl Solution {
    pub fn min_swaps(s: String) -> i32 {
        let mut extra_close = 0;
        let mut max_close = 0;

        for c in s.chars() {
            if c == '[' {
                extra_close -= 1;
            } else {
                extra_close += 1;
            }

            max_close = max_close.max(extra_close);
        }

        (max_close + 1) / 2
    }
}